### ISSUE TITLE
Colour Temperature Controls

### ISSUE THREAD

**Issue by PorscheMonty** [chocolate-doom/chocolate-doom#1139]
Hello there.

Back in the day we used CRT monitors which were mostly set to operate at 9300K by default, and it's quite clear that the people behind Doom's artwork were not the exception.

When Doom is played at 6500k, the industry's standard in 2019, colours will inevitably be off, though this is hardly noticed by the common player today because LCD colour reproduction is still arse, but the time is approaching when we'll be dealing with flatpanel technologies that may very well rival anything we've seen in the CRT world (rooting for Crystal LED myself), and believe me, Doom played on a CRT at 6500k is an UNGODLY mess of reds and ochres; it's got to be played at 9300k (or something similar on the bluish side), period.

Now, having colour temperature controls would be very handy, specially when stuck with hardware/ drivers lacking such functionality, as it is the case right now with my laptop, a 10-something year old VAIO.  Sure, third party software and a number of tricks can be used to get around it, but I believe this feature request to be within the limits of what's reasonable and commendable for a project like CD.

Again, this is not an issue right now (6500k and 9300k look different but that's pretty much it; neither strikes me as substantially better on a plain LCD, even a modern, expensive one), but chances are CRT-like colour reproduction will be viable in the near future, so we better be ready.

Thanks for your time and sorry for the inconveniences.

**Comment by jmtd** [comment]
That's a really interesting point. Do you happen to know what the colour temperature was of displays used with NeXT machines too?

**Comment by fabiangreffrath** [comment]
I guess this requires a GL shader for implementation? Or could we simply project a light blue transparent pane over the rendered texture (of course not, but has anyone ever tried how it would look like)?

**Comment by turol** [comment]
Adding a GL shader would require replacement of the blitting code which is quite a lot of work. However adjusting the palettes should do the same trick.

in either case, how exactly do the colors need to be adjusted? Is there a standard algorithm? Quick googling found some blog posts but nothing that looks official.

**Comment by fabiangreffrath** [comment]
But then, an adjusted palette could get loaded as a PWAD. There are already plenty out there which aim to emulate the "look and feel" of 1990s CRT screens.

**Comment by JNechaevsky** [comment]
Indeed, it's a huge problem, especially if you have worked a lot with graphics, if you know the difference... And once you'll notice this difference, it is possible to accept it, but, heh, it can't be unseen.

As Fabian have said, the most simple way is to modify `PLAYPAL` lump, this will not require any changes in the code, and will work for almost every existing source port. There are two things necessary though:

- Understanding how to work and import palettes in Slade.
- Knowing how to use some kind of graphical editors (Photoshop is a de facto, no idea about free tools).

Sadly, I don't know and can't recommend any existing solutions with custom `PLAYPAL`.

**Comment by PorscheMonty** [comment]
> 
> 
> That's a really interesting point. Do you happen to know what the colour temperature was of displays used with NeXT machines too?

Well, I'd have to guess it was the same as other monitors from that time period. Of course, some of them probably had the ability to switch between different values, but that's something I have no recollection ever seeing in the wild for even industrial-grade hardware older than 20 years. Now If it were an interlaced CRT (and to the best of my knowledge it wasn't), there's a slim chance it may have used something a tad bit higher like 9500K, because that's what standard TV phosphors were known to produce, but it's been a while since I went through manuals and canonical papers so this don't quote me on any of it.

Either way, the artists were using MS-DOS on regular PC's, so we're not really missing out on anything here.

As for a potential implementation, well, that's not up to me, but patching up the palette at runtime (barely a few bytes, nothing major) seems like the way to go, because shaders would make it unusable on platforms where they are not properly supported (thinking my WinXP-OpenBSD rig with GF6 card + legacy drivers) and doing fullscreen color correction may imply a relatively important performance hit depending on how it's done.

**Comment by chungy** [comment]
My 2¢: I don't think this is really an appropriate feature for Chocolate Doom, and I don't think the premise is really all that accurate, it is my recollection that CRTs were generally warmer than LCDs, which is the opposite of your claim in the opening post. Nor do I think there is anything obvious about what color temperatures the graphics were “intended” to be displayed with.

I believe you can achieve the same effect with various other means, including controls on your monitor, your operating system, or PLAYPAL hacking.

**Comment by PorscheMonty** [comment]
> 
> 
>  it is my recollection that CRTs were generally warmer than LCDs, which is the opposite of your claim in the opening post.

When a CRT is calibrated towards 6500k, it's going to look "warmer" than LCD's operating under the exact same colour temperature simply because LCD's cannot reproduce colours anywhere near as accurately,  saturation particularly coming heavily into play. That alone  could account for what you allegedly remember, yet it still wouldn't apply to your everyday monitor from the early 90's, which was in fact, calibrated for 9300k (I still have a bunch of them laying around and they all look "cooler" on factory defaults, and none save 2 actually have colour temperature controls, but they were manufactured back in 2003 and 2005 respectively)

> Nor do I think there is anything obvious about what color temperatures the graphics were “intended” to be displayed with.

It's not that they put any thought into it. Colour temperature wasn't something that people really begin to grasp till 6500k was standardized, which thing would not happen till years after that Doom's release; A cooler colour temperature was the norm back then so they worked around it.

Again, it's not going to be obvious unless you're working with a real CRT or a competent flatpanel technology.

> I believe you can achieve the same effect with various other means, including controls on your monitor, your operating system, or PLAYPAL hacking.

That was addressed in my original post, but it can be pretty inconvenient or just plain nonviable for certain setup's.

**Comment by jmtd** [comment]
The problem with solving this with a PLAYPAL lump is it would either override, or be overridden, by a PWAD's custom PLAYPAL lump. If we were to implement this, it would have to be a runtime adjustment to whatever PLAYPAL lump was selected from input WADs.

**Comment by PorscheMonty** [comment]
Btw, there's a variety of ways we can calculate the deviations including a couple of formulas, but I myself ran the Colour Axis Transformation tool from Color Space 1.1.1 (couleour.org) on the whole 10752 bytes of PLAYPAL and it worked out just fine. The GIMP does provide a similar yet vastly less flexible option, but unlike CS111, 9300k is one of the predefined illuminants so you don't even have to type anything in. Obviously, this is just for testing given that, if at all implemented, it would have to be done at runtime as you yourself pointed out.

**Comment by fabiangreffrath** [comment]
Could you show some results, e.g. screen shots, please?

**Comment by PorscheMonty** [comment]
![9300k](https://user-images.githubusercontent.com/10228056/53074813-a4495280-34e3-11e9-9928-6748120f38b6.PNG)
[9300K.zip](https://github.com/chocolate-doom/chocolate-doom/files/2883540/9300K.zip)

Sure thing.
There's this APNG, or you may try the attached wad.

**Comment by fragglet** [comment]
Have you tried pal_plus? https://www.doomworld.com/idgames/graphics/pal_plus

**Comment by PorscheMonty** [comment]
A few things:

1. Yes I have, and it made everything look twice as washed out and lifeless.

2. You cannot improve Doom's look on standard LCD technology; you'd be trading one thing for the other and in this particular case, colours had to make room for a fake contrast that doesn't actually work,

3. The white point wasn't even altered at all so whatever the author of this abomination was hoping for, colour temperature correction it did not.

4. Let's not forget this isn't about getting Doom to overcome the limitations of current LCD displays, but to allow for a rather a minor (yet very noticeable and necessary) correction for CRT's or superior displays that are yet to penetrate the desktop computer market so users don't have to mess around with the  monitor's settings or third party software.

**Comment by fabiangreffrath** [comment]
> Have you tried pal_plus? https://www.doomworld.com/idgames/graphics/pal_plus

I prefer the Jovian Palette: https://www.doomworld.com/idgames/graphics/jovpal

**Comment by jmtd** [comment]
@PorscheMonty , what do you think of the Jovian Palette: does that make an appropriate colour temperature compensation?

Unfortunately the copyright on that PWAD is restrictive and we couldn't directly derive from it if we wanted to. The author also does not document the methadology behind how they came up with the new palette.

What we would probably need to do this in code is effectively a (pure) function to run over the palette that is determined at runtime after PWAD generation.

**Comment by PorscheMonty** [comment]
> 
> 
> @PorscheMonty , what do you think of the Jovian Palette: does that make an appropriate colour temperature compensation?

I'm afraid it doesn't even touch it (not that the author says it does anyways).

Possibly the easiest way to go about this would be to generate the new temperature assuming 6500K as the starting point, and then apply it directly to PLAYPAL.

https://academo.org/demos/colour-temperature-relationship/ can help you with the generation part (we don't need the whole thing, just a few reasonable, pre-calculated values like 9300K and 9500K), and once you have it, here's how you may apply it (there are different methods but this likely the most convenient one due to it not involving colourspace conversion from RGB):

1. Lower the output levels for the 2 channels individually according to the generated RGB values.
2. Lower the input levels of the sum of the three channels (R+G+B) down to highest individual output value that is lower than 255.

E.g. 

1. Desired colour temperature = 9300K (207.221.255).
2. A white pixel becomes 207.221.255 by changing the output levels.
3. This 207.221.255 pixel becomes 239.255.255 after input values are corrected for brightness.
